dev/Kozani/Common/KozaniDvcProtocol.cpp

```diff
@@ -0,0 +1,116 @@
+        THROW_HR_IF(E_UNEXPECTED, rawPDU.empty());
+
+        return std::move(rawPDU);
+    }
+
+    std::string CreateConnectionAckPDU(PCSTR connectionId, UINT64 activityId)
+    {
+        Dvc::ConnectionAck ackMessage;
+        ackMessage.set_connection_id(connectionId);
+        return CreatePDU(activityId, Dvc::ProtocolDataUnit::ConnectionAck, ackMessage.SerializeAsString());
+    }
+
+    std::string SerializeActivatedEventArgs(winrt::Windows::ApplicationModel::Activation::IActivatedEventArgs& args)
+    {
+        switch (args.Kind())
+        {
+        case winrt::Windows::ApplicationModel::Activation::ActivationKind::Launch:
+            auto specificArgs{ args.as<winrt::Windows::ApplicationModel::Activation::LaunchActivatedEventArgs>() };
+            if (!specificArgs.Arguments().empty())
+            {
+                const std::string argsUtf8{ ::Microsoft::Utf8::ToUtf8(specificArgs.Arguments().c_str()) };
+                Dvc::LaunchActivationArgs launchArgs;
+                launchArgs.set_arguments(std::move(argsUtf8));
+                return std::move(launchArgs.SerializeAsString());
+            }
+
+            break;
+        }
+        return std::string();
+    }
+
+    std::string CreateActivateAppRequestPDU(
+        UINT64 activityId,
+        PCWSTR appUserModelId,
+        winrt::Windows::ApplicationModel::Activation::ActivationKind activationKind,
+        winrt::Windows::ApplicationModel::Activation::IActivatedEventArgs& args)
+    {
+        Dvc::ActivateAppRequest activateAppRequest;
+        activateAppRequest.set_activation_kind(static_cast<Dvc::ActivationKind>(activationKind));
+
+        const std::string appUserModelIdUtf8{ ::Microsoft::Utf8::ToUtf8(appUserModelId) };
+        activateAppRequest.set_app_user_model_id(std::move(appUserModelIdUtf8));
+
+        if (args)
+        {
+            activateAppRequest.set_arguments(SerializeActivatedEventArgs(args));
+        }
+
+        return CreatePDU(activityId, Dvc::ProtocolDataUnit::ActivateAppRequest, activateAppRequest.SerializeAsString());
+    }
+
+    std::string CreateActivateAppResultPDU(
+        UINT64 activityId, 
+        HRESULT hr, 
+        DWORD appProcessId, 
+        bool isNewInstance,
+        _In_ const std::string& errorMessage)
+    {
+        Dvc::ActivateAppResult activateAppResult;
+        activateAppResult.set_hresult(hr);
+        if (SUCCEEDED(hr))
+        {
+            activateAppResult.set_process_id(appProcessId);
+            activateAppResult.set_is_new_instance(isNewInstance);
+        }
+        else if (!errorMessage.empty())
+        {
+            activateAppResult.set_error_message(errorMessage);
+        }
+
+        return CreatePDU(activityId, Dvc::ProtocolDataUnit::ActivateAppResult, activateAppResult.SerializeAsString());
+    }
+
+    std::string CreateAppTerminationNoticePDU(UINT64 activityId)
+    {
+        return CreatePDU(activityId, Dvc::ProtocolDataUnit::AppTerminationNotice);
```

Comment:
We use activity ID to link every app activation and termination. So if you launch 3 instances of the app then there will be 3 activity IDs and you will receive 3 separate AppTerminationNotice messages when all of them are terminated. With the ID, we can unqiuely identify the app instance's process if needed. Since we use "App" over the API surface, we should keep it consistent. It is more like "AppInstance", "App" for short.

Reply:
Ahhh. AppInstance makes sense, just wasn't entirely apparent even after reading thru the code, and "App" is a widely used term for "Application" (starting with AppUserModelID, and widely appearing across Windows' WinRT API surface).

CONSIDER changing names to "AppInstance" to avoid ambiguity.

Reply:
Speaking of...capturing this sort of information in a central location will be helpful to future devs.

RECOMMEND: Add `specs\Kozani\KozaniDesignSpec.md` and mention AppInstance, plus PDU and any other key concepts. A full spec would be nice <sic> but at least a README level of minimal but key concepts would be welcome.